### ISSUE TITLE
✨ Allow simple way to create custom block sorting

### ISSUE THREAD

**Issue by MiWeiss** [sciunto-org/python-bibtexparser#480]
We'd want to facilitate users sorting blocks according to their needs. 

They easiest and most general way to do this would probably be to create a new abstract class `SortBlocksMiddleware` with the following properties:

- The abstract class should have just one unimplemented method, and it should have the signature `compare(block1, block2) -> int`, following the standard convention that any returned value > 0 indicates that block2 should come anywhere after block1, any value <0 the opposit and a return value of 0 sais that any order of the two should be fine. 
- The docstring of the `compare` method should list the standard conventions for such a method (the conditions above, but also that it's transitive, pure, ...)
- Our own `SortBlocksByTypeAndKeyMiddleware` should extend/implement `SortBlocksMiddleware`
- The documentation should provide a simple MWE

Location of the code: https://github.com/sciunto-org/python-bibtexparser/blob/main/bibtexparser/middlewares/sorting_blocks.py

I am currently not working on this, so feel free to comment below if you want to take over.

**Comment by mirhahn** [comment]
Hi, 

I would like to take a crack at this. I have a code at work that renders BibTeX into HTML and I ran into the issue of having to code custom sorting for the `bibtexparser-v2` port as well (I sort by keys made up of multiple fields, including author names and release dates, that are supposed to be specified by the user).

I have a few questions: 

1. Do you have style guidelines aside from those mentioned in this Issue and in `CONTRIBUTING.md`? Do you have a preferred docstring format?
2. How exactly do the `allow_inplace_modification` and `allow_parallel_execution` flags work? Especially for the latter, what do I need to do to accommodate that functionality?
3. What is your targeted minimal Python version? It's not specified in `setup.py` as far as I can see and I like heavily annotating my code with type hints, which can easily slip in dependencies on newer standard libraries if I don't know what to target. I assume everything post-3.8 is okay, right?
4. Do you know of a good way to generate random BibTeX files for testing?

It might take me a bit. Day job takes precedence.

**Comment by MiWeiss** [comment]
Hi @mirhahn 

Awesome, thanks a lot!

> Do you have style guidelines aside from those mentioned in this Issue and in CONTRIBUTING.md? Do you have a preferred docstring format?

No. Go ahead. I personally like google docstyle the best and think it's used most frequently in the repo, but so far we're not enforcing it. Should probably do that eventually, though ;-)

> How exactly do the allow_inplace_modification and allow_parallel_execution flags work? Especially for the latter, what do I need to do to accommodate that functionality?

Well, if your implementation is capable of parallel execution (has no race condition, is threadsafe, ...) just pass "true" to the superclass. Otherwise, pass "false". Note however thata at the moment, parallel execution is not yet supported, thus this is more something to be prepared once this is eventually implemented in the parser. 

> What is your targeted minimal Python version? It's not specified in setup.py as far as I can see and I like heavily annotating my code with type hints, which can easily slip in dependencies on newer standard libraries if I don't know what to target. I assume everything post-3.8 is okay, right?

Any version which is not yet EOL. Thus, I assume we'll drop 3.8 support [later this year](https://devguide.python.org/versions/).

> Do you know of a good way to generate random BibTeX files for testing?

Search for `.bib` files on github or google. This gives you real data, which is better than random ;-) Otherwise, use your favorite LLM (e.g. ChatGPT) , which are pretty good at generating reasonable test data, although in my experience often lack corner cases and diversity.

**Comment by MiWeiss** [comment]
@mirhahn A assume you're not working on this anymore and I can unassign?

**Comment by mirhahn** [comment]
Yeah, I'm really sorry. I dramatically misjudged my capacity to allocate time for side projects and then lost track. I'm very sorry for slowing this down.

If it helps, I would suggest going for a key-type system instead of a comparator system (similar to how the `key` argument of Python's own `sort` works). Supporting both simultaneously is difficult and tends to rely on somewhat flimsy uses of the `inspect` module. Comparators can be turned into key object generators using `functools.cmp_to_key` but there is no vice-versa conversion.

A key-type system allows users to use simple lambdas like `lambda x: x["title"]` to sort by specific properties or even `lambda x: (x["year"], x["title"])` for hierarchical sorting. Writing a comparator function can be tedious for more complex sorting criteria.

Unfortunately, I got sidetracked trying to accommodate both, so I ended up completing neither. Again, I am very sorry for this.

01.06.2026 08:04:08 Michael Weiss ***@***.***>:
